### PR TITLE
Merge [GC] Code Refactoring, ZAttachedArray, ZSafeDelete

### PR DESCRIPTION
**Pull request by tanghaoth90**
8219633: ZGC: Rename ZPageSizeMin to ZGranuleSize
8219634: ZGC: Rename ZAddressRangeMap to ZGranuleMap
8220587: ZGC: Break out forwarding information from ZPage
8220588: ZGC: Convert ZRelocationSet to hold ZForwardings instead of ZPages
8220589: ZGC: Remove superfluous ZPageTableEntry
8220590: ZGC: Remove ZPages from ZPageTable when freed
8220591: ZGC: Don't delay reclaimation of ZVirtualMemory
8220592: ZGC: Move destruction of detached ZPages into ZPageAllocator
8220593: ZGC: Remove superfluous ZPage::is_detached()
8220594: ZGC: Remove superfluous ZPage::is_active()
8220595: ZGC: Introduce ZAttachedArray
8220595: ZGC: Introduce ZAttachedArray
8220597: ZGC: Convert ZForwarding to use ZAttachedArray
8220599: ZGC: Introduce ZSafeDelete
8219816: Add IsArray/RemoveExtent type traits utilities
8220600: ZGC: Delete ZPages using ZSafeDelete

@mmyxym @linade @leveretconey

### CODE REVIEW

**Review by mmyxym — APPROVED** [review]

File: src/hotspot/os_cpu/linux_x86/gc/z/zGlobals_linux_x86.hpp [hunk 1]
```diff
@@ -74,7 +74,7 @@
 //  * 63-47 Fixed (17-bits, always zero)
 //
 
-const size_t    ZPlatformPageSizeSmallShift   = 21; // 2M
+const size_t    ZPlatformGranuleSizeShift      = 21; // 2M
```
Comment (linade) [inline src/hotspot/os_cpu/linux_x86/gc/z/zGlobals_linux_x86.hpp:77]: extra space

File: test/hotspot/jtreg/ProblemList-zgc.txt [hunk 2]
```diff
@@ -0,0 +1,42 @@
...
+serviceability/sa/ClhsdbJstack.java                           8220624   generic-all
+serviceability/sa/ClhsdbPrintAs.java                          8220624   generic-all
+serviceability/sa/ClhsdbPstack.java                           8220624   generic-all
+serviceability/sa/ClhsdbSource.java                           8220624   generic-all
+serviceability/sa/TestClhsdbJstackLock.java                   8220624   generic-all
+serviceability/sa/TestHeapDumpForInvokeDynamic.java           8220624   generic-all
+serviceability/sa/TestHeapDumpForLargeArray.java              8220624   generic-all
+serviceability/sa/TestUniverse.java                           8220624   generic-all
+serviceability/sa/TestJmapCore.java                           8220624   generic-all
+serviceability/sa/TestJmapCoreMetaspace.java                  8219443   generic-all
```
Comment (linade) [inline test/hotspot/jtreg/ProblemList-zgc.txt:42]: why is here an extra test?

Comment (tanghaoth90) [reply test/hotspot/jtreg/ProblemList-zgc.txt]: I will remove the last line

File: src/hotspot/os_cpu/linux_x86/gc/z/zGlobals_linux_x86.hpp (hunk 1 above)
Comment (tanghaoth90) [reply src/hotspot/os_cpu/linux_x86/gc/z/zGlobals_linux_x86.hpp]: resolved

**Review by linade — APPROVED** [review]

**Review by leveretconey — APPROVED** [review]

**Pull request merged**